docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx

```diff
@@ -0,0 +1,240 @@
+Click on a leaf node to open the side-panel. You'll see the following stub provided in the code editor:
+
+```javascript
+class LeafNode extends TreeNode {
+  public override async execute(rg): Promise<NodeStatus> {
+    return NodeStatus.SUCCESS;
+  }
+}
+```
+
+:::info
+All of your node's code must be defined within the `LeafNode` class, and the `execute` method must be defined as your entrypoint into the node.
+You may encounter errors loading your AI into a game if these are not properly defined when you save your tree.
+:::
+
+You can either manually type code into the editor (see the [PlayTest Bots API](./playtest-bots)), or you can have GPT do the heavy lifting for you.
+To generate code using GPT, describe what you want this node to do in the "GPT Prompt" field then click the "Generate Code" button.
+The editor will be disabled as code appears, and you can modify the output once it's finished.
+
+:::caution
+Generating Code will overwrite anything currently in the editor. It's a good idea to save your tree before trying out new prompts.
+:::
+
+You may have many decision-making points and available actions depending on the complexity of your game,
+so let's start with something simple.
+Let's assume that an "Enemy" ObjectType is defined within our game and that the bot's character has the
+"ApproachEntity" and "AttackEntity" actions.
+
+If we enter the following prompt:
+```text
+If there is an Enemy within 10 square meters then approach them and store a reference to them
+```
+
+then GPT should output code similar to this:
+```typescript
+class LeafNode extends TreeNode {
+  public override async execute(rg): Promise<NodeStatus> {
+    const bot = rg.getBot();
+    const enemy = await rg.findNearestEntity("Enemy", bot.position, (entity) => {
+      return rg.MathFunctions.distanceSq(bot.position, entity.position) <= 10;
+    });
+
+    if (enemy) {
+      rg.performAction("ApproachEntity", { targetId: enemy.id });
+      this.setData("enemy", enemy);
+      return NodeStatus.SUCCESS;
+    }
+
+    return NodeStatus.FAILURE;
+  }
+}
+```
+
+We can then enter another prompt in a separate node to complete our goal:
+```text
+Attack the stored enemy
+```
+```typescript
+class LeafNode extends TreeNode {
+  public override async execute(rg): Promise<NodeStatus> {
+    const enemy = this.getData<RGEnemyState>("enemy");
+    if (!enemy) {
+      return NodeStatus.FAILURE;
+    }
+    rg.performAction("AttackEntity", { targetId: enemy.id });
+    return NodeStatus.SUCCESS;
+  }
+}
+```
+
+The above outputs are good examples of how properly-formatted code should look when you're finished generating or writing it.
+
+You may notice that these examples invoke the `getData` and `setData` methods.
+These can be used to share data between nodes, and it's a good practice to specify a data type for `getData` to take advantage of the editor's Intellisense.
+
+:::note
+Classes to represent available state variables for entities are coming soon
+:::
```

Comment:
Looking at the code, I think this is a little more accurate (I noticed that the data appears to be shared across ticks too and that was initially surprising to me, but I think it's useful):

```suggestion
These can be used to share data between nodes and across each tick of your bot.

:::note
We recommend you specify a data type in `getData` so you can take full advantage of autocomplete in the editor.
:::
```

And again, no need for a "Coming Soon" here I think. We don't know when "soon" is, and when "soon" does come, it'd be better if we don't have to update this note.

Reply:
This is the only comment that I disagree with. I think it's going to be very confusing for someone to read "make sure you specify a type!" in the docs when the relevant types don't exist yet. We need some indication that we recommend the use of types where they _can_ be applied, but that we're still working on implementing types to represent your states and actions

Reply:
Maybe "under development" is a better phrase since you're right that we don't know how _soon_ "soon" is

Reply:
> I think it's going to be very confusing for someone to read "make sure you specify a type!" in the docs when the relevant types don't exist yet.

A very fair point, though perhaps one that argues for just omitting the "make sure you specify a type" comment until we actually have the type infrastructure in place?

Either way, I'm OK with it. I'm certainly not significantly opposed to referencing future work. I think in a previous iteration (before I saved the comment) I had something like this:

```suggestion
These can be used to share data between nodes and across each tick of your bot.

:::note
We recommend you specify a data type in `getData` so you can take full advantage of autocomplete in the editor. The types available in the editor are limited, but we're continually adding more!
:::
```